lib/hooks.dart

```diff
@@ -79,7 +65,10 @@ class StateHook<T> {
 /// ```
 ///
 /// Learn more: <https://reactjs.org/docs/hooks-state.html>.
-StateHook<T> useState<T>(T initialValue) => StateHook(initialValue);
+StateHook<T> useState<T>(T initialValue) {
+  final result = React.useState(initialValue);
+  return StateHook._(result[0] as T, result[1] as void Function(dynamic));
+}
```

Comment:
Should this be

```suggestion
StateHook<T> useState<T>(T initialValue) {
  final result = React.useState(initialValue);
  return StateHook._(result[0] as T, result[1] as void Function(T));
}
```

Or is this not possible because `T` is a dart type, and the function type casting here is representative of a value returned from JS only? If the latter, that might be worth a comment in the code.

Reply:
Yeah unfortunately the runtime type of the function is always going to be `Function(dynamic)` since it comes from the JS, so that would be an invalid cast. I can def add a comment 👍

Reply:
Oh looks like I was wrong there; `void Function(dynamic)` is technically a subtype of `void Function(T)`, so we could have done that. Looks like we actually do that in `ReducerHook`.

However, in this case we do `dynamic` instead of `T`, because this function can be called with either the value or an updater function. So the typing is really `void Function(T|(T Function(T)))`, which in Dart's type system is only expressible as `void Function(dynamic)`.

I added some comments and a shared typedef for this to help make it clearer: https://github.com/Workiva/react-dart/pull/373/commits/7ab1084bf79ac8049b2921b64aa1088113837060